odxtools/cli/_print_utils.py

```diff
@@ -53,31 +63,197 @@ def print_diagnostic_service(
+    table = extract_parameter_tabulation_data(list(service.request.parameters))
+    print(tabulate(table, headers='keys', tablefmt='presto'))
+    print(f"\n   Message format of the request:")
+    service.request.print_message_format(
+        indent=0, allow_unknown_lengths=allow_unknown_bit_lengths, print_objects=print_objects)
 
-            print(f"  Message format of a negative response:")
-            resp.print_message_format(indent=3, allow_unknown_lengths=allow_unknown_bit_lengths)
+    # Positive Response
+    print(f"\n  [yellow]Positive Response Properties[/yellow]:")
+    print(f"   Number of Positive Responses: {len(service.positive_responses)}")
+    print(f"   Positive Responses: {service.positive_responses}")
+    if len(service.positive_responses) == 1:
+        resp = service.positive_responses[0]
+        print(f"   Service Parameters: {resp.parameters}\n")
+        table = extract_parameter_tabulation_data(list(resp.parameters))
+        print(tabulate(table, headers='keys', tablefmt='presto'))
+        print(f"\n   Message format of the positive response:")
+        resp.print_message_format(
+            indent=0, allow_unknown_lengths=allow_unknown_bit_lengths, print_objects=print_objects)
+
+    # Negative Response
+    print(f"\n  [yellow]Negative Response Properties[/yellow]:")
+    print(f"   Number of Negative Responses: {len(service.negative_responses)}")
+    print(f"   Negative Responses: {service.negative_responses}")
+    if len(service.negative_responses) == 1:
+        resp = service.negative_responses[0]
+        print(f"   Service Parameters: {resp.parameters}\n")
+        table = extract_parameter_tabulation_data(list(resp.parameters))
+        print(tabulate(table, headers='keys', tablefmt='presto'))
+        print(f"\n   Message format of a negative response:")
+        resp.print_message_format(
+            indent=0, allow_unknown_lengths=allow_unknown_bit_lengths, print_objects=print_objects)
+
+    print("\n")
 
     if (service.positive_responses and
             len(service.positive_responses) > 1) or (service.negative_responses and
                                                      len(service.negative_responses) > 1):
         # Does this ever happen?
         raise NotImplementedError(
             f"The diagnostic service {service.odx_id} offers more than one response!")
+
+
+def extract_service_tabulation_data(services: List[DiagService]) -> Dict[str, Any]:
+    # extracts data of diagnostic services into Dictionary which can be printed by tabulate module
+
+    name = []
+    semantic = []
+    request: List[Optional[str]] = []
+
+    for service in services:
+        assert isinstance(service, DiagService)
+        name.append(service.short_name)
+        semantic.append(service.semantic)
+        if not getattr(service.request, "required_parameters", None):
+            if len(service().hex().upper()) > 24:
+                request.append('0x' + service().hex().upper()[:24] + '...')
```

Comment:
Maybe a function that takes bytes as input, and return truncated data or full data as string? same logic in print_service_parameters

Reply:
new function `format_request_string` was introduced.